src/core/net/SRP_CODER.md

```diff
@@ -0,0 +1,284 @@
+- `110 <5-bit code>`: Commonly used constant labels. Codes are:
+  - `0`: `_udp`
+  - `1`: `_tcp`
+  - `2`: `_matter`
+  - `3`: `_matterc`
+  - `4`: `_matterd`
+  - `5`: `_hap`
+- `111 <5-bit code>`: Commonly used generative patterns. Codes are:
+  - `0`: Single 64-bit value converted to a 16-char hexadecimal string using capital letters. The 64-bit value is encoded in binary format (big-endian). This pattern is used for target host name in matter.
+  - `1`: Two 64-bit values each expressed as a 16-char hexadecimal string, using capital letters, separated by a hyphen `-`. The two 64-bit values are encoded in binary format (8 bytes big-endian order). This pattern is used for service instance name in matter.
+  - `2`: Subtype label as `_<char><hhhh>` where a single char is encoded first, followed by 8 byte encoded 64-bit value for `<hhhh>` which is converted to a 16-char hexadecimal string, using capital letters.
+  - `3` : Subtype label as `_<char><hhhh>`, similar to previous case but instead of encoding the 8-byte value directly, an offset to an earlier instance of the same byte sequence is encoded in the message. The offset is encoded using compact `uint` format.
+
+A DNS name is encoded as a sequence of labels, ending with an empty label encoded as a single `0x00` byte. DNS names do not include the domain name.
+
+#### Examples
+
+`_service._udp` is encoded using `10` bytes (vs normal DNS encoding of `15`):
+
+```
+  +----+----+----+----+----+----+----+----+----+----+
+  | 47 | s  | e  | r  | v  | i  | c  | e  | C0 | 00 |
+  +----+----+----+----+----+----+----+----+----+----+
+```
+
+- `0x47` (`0b0100_0111`) - label starting with `_` with `len=7`.
+- `0xC0` (`0b110_0000`) - Commonly used constant label - code 0 for `_udp`.
+- `0x00` end.
+
+`2906C908D115D362-8FC7772401CD0696` (matter service instance label example) encoded using `17` bytes (vs uncompressed DNS encoding of `34` bytes).
+
+- `0xE1` (`0b1110_0001`) - Commonly used generative patterns (code 1)
+- `0x29` `0x06` `0xC9` `0x08` `0xD1` `0x15` `0xD3` `0x62` (first 64-bit value)
+- `0x8F` `0xC7` `0x77` `0x24` `0x01` `0xCD` `0x06` `0x96` (second 64-bit value)
+
+## Message Format
+
+- [Header Block](#header-block)
+- Zero or more
+  - [Add Service Block](#add-service-block) or [Remove Service Block](#remove-service-block)
+- [Host Block](#host-block)
+- [Footer Block](#footer-block)
+
+## Header Block
+
+### Header Dispatch
+
+```
+    0      1     2     3     4     5     6     7
+  +-----+-----+-----+-----+-----+-----+-----+-----+
+  |  0  |  0  |  1  |  1  |  0  |  0  |  Z  |  T  |
+  +-----+-----+-----+-----+-----+-----+-----+-----+
+```
+
+- The initial bits in dispatch byte are fixed to differentiate it from the general DNS header used in SRP updates.
+- `Z` flag (Zone)
+  - `0`: Zone name is elided - use the default `default.service.arpa`.
```

Comment:
This makes sense if you are directly constructing the message, but won't work if you are compressing a message, because the first name in the message will probably be default.service.arpa.

Reply:
Added some clarification on this in the new push. Basically, in the encoded message, service names and host names, etc., always omit the domain name and are encoded as one or more labels (terminating with the empty label `0x00`). Decoder will reconstruct the full name.

Reply:
I think part of the problem is that you're over-using the term "domain name" here. Every DNS name is a domain name, so this is confusing to me because I'm so steeped in DNS minutiae. RFC 2136 uses the term "zone name," and you've used it correctly here, but I've noted three places where you use "domain name" when I suspect "zone name" would be clearer.